### PR TITLE
SPV Updates

### PR DESCRIPTION
**Pull request by nodech**
Add reorg tests for Chain Full and SPV.
  - Add chain full/spv reorg test.
  - Port: https://github.com/bcoin-org/bcoin/pull/755 for SPV Full sync
    - SPV reorg will now emit `reset`.
  - fullnode/spvnode/chain will now include `fork` entry in the `reorganize` event after `tip` and `comptetitor`.
  - Check `mtp` and `time-too-new` checks to the SPV Chain.

### CODE REVIEW

File: test/chain-reorg-test.js [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+          const job = await cpu.createJob(tip1);
+          job.refresh();
+
+          const blk = await job.mineAsync();
+          const hash = blk.hash();
+
+          assert(await chain.add(blk));
+
+          tip1 = await chain.getEntry(hash);
+          // last one.
```
Comment (pinheadmz) [inline test/chain-reorg-test.js:230]: "last one" ?

**Review by pinheadmz — COMMENTED** [review]
LGTM just have a few questions and nits.
If you have time, you could also incorporate https://github.com/bcoin-org/bcoin/pull/755 which should fix all the stuck SPV waiting for new blocks crap and make your tests a lot cleaner.

In addition, I wonder if youd be interested in breaking up the two long test blocks into separate `describe` for full and SPV and then breaking up the `{}` blocks as `it`s

**Comment by pinheadmz** [comment]
Synced SPV node from this branch in < 30 minutes. No reorgs while syncing but hey, nice performance.

File: test/chain-test.js [hunk 2]
```diff
@@ -1,1064 +1,425 @@
...
-  try {
-    entry = await chain.add(block, flags);
-  } catch (e) {
-    assert.strictEqual(e.type, 'VerifyError');
-    return e.reason;
+function addBlock(chain, block, wallet) {
+  if (chain.options.spv) {
+    const flags = chainCommon.flags.VERIFY_NONE;
+    const mblock = MerkleBlock.fromBlock(block, wallet.filter);
+    return chain.add(mblock, flags);
```
[the remark this answers is not in the document]
Comment (nodech) [reply test/chain-test.js:24]: This just mimics the SPV behaviour.

**Review by pinheadmz — APPROVED** [review]
ACK 55138329613eec257dff0356f681820f74168e94
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 55138329613eec257dff0356f681820f74168e94
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmKCnkEACgkQ5+KYS2KJ
yTrEKA/8DpLKGZf5uIJjBwNhOyU8blxAhNF10zujNrbA8dtsrO5Rf6tOQWAm91nt
ryQftvzFyZPRj4i1Zq3sd/NJnS4CGBThlegretU47LdVDrOCjw5Z4o+AYIVd79es
rA2qsV2X39KjgY52zqgtvn5zsIvfkghzOp1vPx9QFHQutt2au364LEkId5LI8vsZ
laJRkNrlwm71Wk4pRnG+TaqB5V3g5VfB0TT/2OV912pkr/L7ze5xjseVcLk6fplK
J5PnWYnMyCNPJ4UhsP7UhHW2ShSzHvfWP1WPFg4e8zvKu0Pe9962/KgYfAwhH+23
w2I5KvhRoVwGwP0Y7hz7RUvcQ9iDvU77dEcjorRvv95lK9/LBLORthDHUaJgHEY2
5P75MMWm6eQi5gDInO3yXrtg7dTLbif5aI3vVc0kq3nyDdAbpa9/TzDDper784pV
AuIuIzuO1zYSY7YP2NxedKreAlVqkVlW1fzN1s/F3L7JFC8CWu91FGnGFTsMZ3q0
Ma0YBkzcriLuAy0e6UycSTm3cehlT/JMZxICnJTbiqVffhhg+bJldfx7kQ9DFyZe
pP2At9OcyoFLAEK6mJBi0a4/yBm/CTJ2yB5pnMsIsQ8+MKlo/cmatr4GxGxdhvJi
SdLQODYBpWbk+vn324z8xutUYWjNYXUIvepBnfv8xK1F8QCJptQ=
=0rME
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Pull request merged**